File: school-finder-scraper/utils/rate_limiter.py
import time

class RateLimiter:
    """
    Klasa do ograniczania liczby zapytań w jednostce czasu
    """
    def __init__(self, requests_per_minute):
        """
        Inicjalizacja limitera
        
        Args:
            requests_per_minute (int): Maksymalna liczba zapytań na minutę
        """
        self.requests_per_minute = max(1, requests_per_minute)
        self.interval = 60.0 / self.requests_per_minute  # Interwał w sekundach
        self.last_request_time = 0
    
    def wait(self):
        """
        Oczekiwanie na możliwość wykonania kolejnego zapytania
        zgodnie z ustawionym limitem
        """
        current_time = time.time()
        elapsed = current_time - self.last_request_time
        
        # Jeśli od ostatniego zapytania minęło mniej czasu niż interwał,
        # czekamy na upłynięcie interwału
        if elapsed < self.interval:
            sleep_time = self.interval - elapsed
            time.sleep(sleep_time)
        
        # Aktualizacja czasu ostatniego zapytania
        self.last_request_time = time.time()

File: src/lib/scraper/utils/rate-limiter.ts
/**
 * Rate Limiter dla scrapera
 * 
 * Klasa do zarządzania częstotliwością zapytań HTTP
 */

export class RateLimiter {
  private requestsPerMinute: number;
  private interval: number;
  private lastRequestTime: number;
  
  /**
   * Konstruktor
   * @param requestsPerMinute Maksymalna liczba zapytań na minutę
   */
  constructor(requestsPerMinute: number = 10) {
    this.requestsPerMinute = requestsPerMinute;
    this.interval = 60 * 1000 / requestsPerMinute; // Interwał w milisekundach
    this.lastRequestTime = 0;
  }
  
  /**
   * Oczekiwanie na możliwość wykonania kolejnego zapytania
   */
  async wait(): Promise<void> {
    const now = Date.now();
    const timeElapsed = now - this.lastRequestTime;
    
    if (timeElapsed < this.interval) {
      const delay = this.interval - timeElapsed;
      await new Promise(resolve => setTimeout(resolve, delay));
    }
    
    this.lastRequestTime = Date.now();
  }
}